File: src/assignments/week_2/machine-learning-ex1/ex1/gradientDescent.m
function [theta, J_history] = gradientDescent(X, y, theta, alpha, num_iters)
%GRADIENTDESCENT Performs gradient descent to learn theta
%   theta = GRADIENTDESCENT(X, y, theta, alpha, num_iters) updates theta by 
%   taking num_iters gradient steps with learning rate alpha

% Initialize some useful values
m = length(y); % number of training examples
J_history = zeros(num_iters, 1);

for iter = 1:num_iters
    %we do linear regression with identity as an activation function f(x) =
    %x
    prediction = X * theta;
    delta = (1/m) * X' * (prediction - y);
    theta = theta - alpha * delta;
    % Save the cost J in every iteration    
    J_history(iter) = computeCost(X, y, theta);
end

end

File: algorithms/assignments/python/week_2/ex1/gradientDescent.py
import numpy as np
import computeCost as costModule

def gradientDescent(X, y, theta, alpha, num_iters):
    """Performs gradient descent to learn theta
    theta = GRADIENTDESCENT(X, y, theta, alpha, num_iters) updates theta by 
    taking num_iters gradient steps with learning rate alpha """
    
    m = y.size #number of training examples
    J_history = np.zeros((num_iters, 1))
    
    print("theta before for", theta.shape)
    
    #train 
    for iter in range(num_iters):
         #do linear regression with identity (f(x) = x) as an activation function
         prediction = np.dot(X, theta)
         errors = np.subtract(prediction, y)
         delta = (1.0/m) * np.dot(X.T, errors)
         #update weight
         theta = theta - alpha * delta
         #save the cost J in every iteration
         J_history[iter] = costModule.computeCost(X, y, theta)
     
    return J_history, theta
    